### PR TITLE
Add Conditional SMC (C-SMC) inference

### PR DESCRIPTION
**Pull request by ClementeP**
Implements Conditional SMC per Andrieu, Doucet & Holenstein (2010), §4.3. Slot 0 of the particle population is a *retained particle* — extended deterministically along a pre-specified token sequence and exempt from resampling; the remaining `n_particles - 1` slots evolve as in standard SMC.

### New file: `genlm/control/sampler/csmc.py`

- **`csmc_standard(model, n_particles, ess_threshold=0.5)`** — SMC loop with the modified resampling rule: slot 0 survives unchanged; the remaining `N-1` slots are drawn multinomially from all `N` normalized weights. Sets and re-establishes the `is_retained` slot tag after every resample, so retainedness is a *slot* property — a free slot that inherits slot 0's trajectory via multinomial resampling is correctly marked free going forward.

- **`RetainedTokenSampler(base, retained_sequence)`** — transparent wrapper around any `TokenSampler`. When its host particle has `is_retained == True` and the retained sequence isn't exhausted, forces the next token by passing a constant `draw` callable to the base sampler's `sample()`. Weights are computed by the base sampler at the forced token, so weight semantics match standard SMC exactly. Otherwise a no-op delegate.

### Modified: `genlm/control/sampler/sequence.py`

- `SMC.__call__` gains an optional `retained_sequence` kwarg. When `None`, the standard-SMC code path runs byte-identical to before. When provided, the unit sampler is wrapped in `RetainedTokenSampler` and the loop dispatches to `csmc_standard` instead of `smc_standard`. No new top-level class — the existing `SMC` handles both.

### Tests: `tests/sampler/test_csmc.py` (13 tests, all passing)

- 2 `_tag_slots` invariants — slot 0 retained, others free; deepcopy-inherited tags get cleared.
- 5 `csmc_standard` loop tests — termination, slot-0 history is "retained" throughout, slot-j tag at termination, M=1 edge case, input validation.
- 4 `RetainedTokenSampler` tests — forcing, passthrough when not retained, passthrough on exhaustion, `retained_idx` advance discipline.
- 2 `SMC.__call__` integration tests — with `retained_sequence`, and regression check without (existing path still works).

Existing `tests/sampler/test_seq_sampler.py` (7 tests) unchanged and passing — no regression on the standard SMC path.

### Independence

This PR does not depend on #(prompt_ids contextvar PR). C-SMC works without the prompt fix because the trainer-level call pattern is sequential per question. The prompt fix unlocks `asyncio.gather`'d parallel batched C-SMC at the trainer level (in `genlm-latent`), which is a separate change.

### Reference

Andrieu, Doucet & Holenstein (2010), "Particle Markov chain Monte Carlo methods", JRSS B 72(3):269–342. The C-SMC kernel is defined in §4.3; its invariance for fixed θ is established there (Theorem 5(a) extends it to the full Particle Gibbs joint chain on (θ, x)).

### CODE REVIEW

File: docs/csmc.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# Conditional SMC
+
+Conditional Sequential Monte Carlo (C-SMC) is a variant of SMC in which one slot in the particle population is *retained* across resampling rounds. This page covers the math, the API, and how to use it.
+
+C-SMC is implemented in [`genlm.control.sampler.csmc`][genlm.control.sampler.csmc] and surfaced through `SMC.__call__(retained_sequence=...)`.
+
+## Why
```
Comment (vicky-xef) [inline docs/csmc.md]: The explanation in md file is really focused on our EM latent variable project. Is it necessary to refer to the EM-style training or relate it to TRICE?  I suggest just explaining what C-SMC does in general and how to use it with our codebase.

File: genlm/control/sampler/csmc.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    _tag_slots(particles)
+
+    while any(not p.done_stepping() for p in particles):
+        for p in particles:
+            p.untwist()
+        await asyncio.gather(
+            *[p.step() for p in particles if not p.done_stepping()]
+        )
+
+        W = np.array([p.weight for p in particles])
```
Comment (vicky-xef) [inline genlm/control/sampler/csmc.py]: Check case where np.all(W==-np.inf) (like we do in smc_standard())

File: genlm/control/sampler/csmc.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            for _ in range(n_particles - 1):
+                idx = np.random.choice(n_particles, p=probs)
+                new_particles.append(copy.deepcopy(particles[idx]))
+            particles = new_particles
+
+            avg_weight = w_sum - np.log(n_particles)
+            for p in particles:
+                p.weight = avg_weight
+
+            _tag_slots(particles)
```
Comment (vicky-xef) [inline genlm/control/sampler/csmc.py]: We have 3 sequential loops above. Can we do some operations in the same loop? For example _tag_slots()?

File: genlm/control/sampler/csmc.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+import copy
+import asyncio
+import numpy as np
+from arsenal.maths import logsumexp
+
+from genlm.control.sampler.token import TokenSampler
+
+
+async def csmc_standard(model, n_particles, ess_threshold=0.5):
```
Comment (vicky-xef) [inline genlm/control/sampler/csmc.py]: I think we should implement csmc_standard in llamppl (just like we have smc_standard.py) and import it from there.

**Review by vicky-xef — CHANGES_REQUESTED** [review]

**Review by vicky-xef — APPROVED** [review]
